pkg/buildkit/worker/pool.go

```diff
@@ -240,17 +233,55 @@ func (p *workerPool) leasePod(ctx context.Context, pod *corev1.Pod) (*corev1.Pod
 	return pod, nil
 }
 
+// removes lease metadata from given pod and adds expiry
+func (p *workerPool) releasePod(ctx context.Context, pod *corev1.Pod) error {
+	pac, err := corev1ac.ExtractPod(pod, fieldManagerName)
+	if err != nil {
+		return fmt.Errorf("cannot extract pod config: %w", err)
+	}
+
+	pac.WithAnnotations(map[string]string{
+		expiryTimeAnnotation: time.Now().Add(p.podMaxIdleTime).Format(time.RFC3339),
+	})
+	delete(pac.Annotations, leasedAnnotation)
+	delete(pac.Annotations, managerIDAnnotation)
+
+	p.log.Info("Applying pod metadata changes", "annotations", pac.Annotations)
+	if pod, err = p.podClient.Apply(ctx, pac, metav1.ApplyOptions{FieldManager: fieldManagerName}); err != nil {
+		return fmt.Errorf("cannot update pod metadata: %w", err)
+	}
+
+	p.triggerReconcile()
+
+	return nil
+}
+
 // builds routable url for buildkit pod with protocol and port
 func (p *workerPool) buildEndpointURL(ctx context.Context, podName string) (string, error) {
 	p.log.Info("Querying for endpoints", "name", p.serviceName, "namespace", p.namespace)
-	endpoints, err := p.endpointsClient.Get(ctx, p.serviceName, metav1.GetOptions{})
-	if err != nil {
-		return "", err
-	}
 
-	hostname, err := p.extractHostname(endpoints, podName)
+	var hostname string
+
+	// sometimes it takes a short period of time for the endpoints record to be updated with the latest list of "ready"
+	// pods, so we may need to retry fetching the resource when kubernetes experiences some lag
+	err := retry.OnError(
```

Comment:
FWIW there's the underlying `wait` function that powers this too - 

https://github.com/dominodatalab/hephaestus/blob/5ad22086fb05a3bafa2a552a4279098cf91d9449/pkg/buildkit/archive/archive.go#L74

Reply:
I saw that and considered using it until I dug into the implementation. Unfortunately, the underlying function will 
- exit the first time I return an err (i.e. i have to discard the errors in order to keep retrying)
- the final error returned after timeout is `ErrWaitTimeout`

`retry.OnError` is a bit more complicated than what I need (i.e. retriable is always true), but it does what i want otherwise (i.e. re-run on error and return the real error after timeout).